NI Traktor Kontrol S2 Mk2: add HID controller mapping

Be-ing opened the pull request:
[Wiki page](https://github.com/mixxxdj/mixxx/wiki/Native-Instruments-Traktor-Kontrol-S2-Mk2)

Be-ing committed:
NI Traktor Kontrol S2 Mk2: add HID controller mapping

Be-ing commented:
I only have access to this device for a few more days so please review it soon.

Be-ing commented:
@ywwg you can copy the effects code from this for the Kontrol S3 mapping. Note there is a subtle issue with the effects lights flickering when toggling between showing/hiding parameters. If you could figure out how to fix that, that would be helpful.

Swiftb0y commented on res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js:
```diff
@@ -0,0 +1,1389 @@
+
+  // When true, packets will not be sent to the controller.
+  // Used when updating multiple LEDs simultaneously.
+  this.freeze_lights = false;
```
Can you further explain what the purpose for this is? Seems to be this is basically a semaphore, but why is it needed exactly?

Be-ing replied:
so multiple LED updates can be sent with a single packet rather than sending intermediate packets

Swiftb0y replied:
So iiuc when this flag is set, changes in HID output are being batched together so they can be sent in a single message later?

Be-ing replied:
yes

Holzhaus replied:
Please use camelCase for variable names if you're renaming this anyway.

Swiftb0y commented on res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js:
```diff
@@ -0,0 +1,1389 @@
+  MessageShort.setCallback("[Channel1]", "!play", this.play);
+  MessageShort.setCallback("[Channel2]", "!play", this.play);
+  MessageShort.setCallback("[Channel1]", "!pad1", this.pad);
+  MessageShort.setCallback("[Channel1]", "!pad2", this.pad);
+  MessageShort.setCallback("[Channel1]", "!pad3", this.pad);
+  MessageShort.setCallback("[Channel1]", "!pad4", this.pad);
+  MessageShort.setCallback("[Channel1]", "!remix_button", this.samplerModeButton);
+  MessageShort.setCallback("[Channel1]", "!flux_button", this.introOutroModeButton);
+  MessageShort.setCallback("[Channel2]", "!pad1", this.pad);
+  MessageShort.setCallback("[Channel2]", "!pad2", this.pad);
+  MessageShort.setCallback("[Channel2]", "!pad3", this.pad);
+  MessageShort.setCallback("[Channel2]", "!pad4", this.pad);
+  MessageShort.setCallback("[Channel2]", "!remix_button", this.samplerModeButton);
+  MessageShort.setCallback("[Channel2]", "!flux_button", this.introOutroModeButton);
+  MessageShort.setCallback("[Channel1]", "!loop_in", this.loopIn);
+  MessageShort.setCallback("[Channel1]", "!loop_out", this.loopOut);
+  MessageShort.setCallback("[Channel2]", "!loop_in", this.loopIn);
+  MessageShort.setCallback("[Channel2]", "!loop_out", this.loopOut);
+  MessageShort.setCallback("[Channel1]", "!load_track", this.loadTrack);
+  MessageShort.setCallback("[Channel2]", "!load_track", this.loadTrack);
+  MessageShort.setCallback("[Channel1]", "!sync_enabled", this.syncEnabled);
+  MessageShort.setCallback("[Channel2]", "!sync_enabled", this.syncEnabled);
+  MessageShort.setCallback("[Channel1]", "!left_encoder_press", this.leftEncoderPress);
+  MessageShort.setCallback("[Channel2]", "!left_encoder_press", this.leftEncoderPress);
+  MessageShort.setCallback("[Channel1]", "!right_encoder_press", this.rightEncoderPress);
+  MessageShort.setCallback("[Channel2]", "!right_encoder_press", this.rightEncoderPress);
+  MessageShort.setCallback("[Channel1]", "!jog_touch", this.jogTouch);
+  MessageShort.setCallback("[Channel2]", "!jog_touch", this.jogTouch);
+  MessageShort.setCallback("[Channel1]", "!jog_wheel", this.jogMove);
+  MessageShort.setCallback("[Channel2]", "!jog_wheel", this.jogMove);
+  
+  MessageShort.setCallback("[EffectRack1_EffectUnit1]", "!effect_focus_button", this.effectFocusButton);
+  MessageShort.setCallback("[EffectRack1_EffectUnit1", "!effectbutton1", this.effectButton);
+  MessageShort.setCallback("[EffectRack1_EffectUnit1", "!effectbutton2", this.effectButton);
+  MessageShort.setCallback("[EffectRack1_EffectUnit1", "!effectbutton3", this.effectButton);
+ 
+  MessageShort.setCallback("[EffectRack1_EffectUnit2]", "!effect_focus_button", this.effectFocusButton);
+  MessageShort.setCallback("[EffectRack1_EffectUnit2", "!effectbutton1", this.effectButton);
+  MessageShort.setCallback("[EffectRack1_EffectUnit2", "!effectbutton2", this.effectButton);
+  MessageShort.setCallback("[EffectRack1_EffectUnit2", "!effectbutton3", this.effectButton);
```
Can you condense this sections by constructing this in a loop?
eg 
```javascript
for (group of ["[Channel1]", "[Channel2]"]) {
      MessageShort.setCallback(group, "!top_encoder_press", this.topEncoderPress);
      ...
}
```

Be-ing replied:
Better yet, I made these `setCallback` calls unnecessary by letting the script provide the callback to `addControl`.

Swiftb0y commented on res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js:
```diff
@@ -0,0 +1,1389 @@
+  MessageShort.setCallback("[Channel2]", "!jog_wheel", this.jogMove);
+  
+  MessageShort.setCallback("[EffectRack1_EffectUnit1]", "!effect_focus_button", this.effectFocusButton);
+  MessageShort.setCallback("[EffectRack1_EffectUnit1", "!effectbutton1", this.effectButton);
```
typo?
```suggestion
  MessageShort.setCallback("[EffectRack1_EffectUnit1]", "!effectbutton1", this.effectButton);
```

Swiftb0y commented on res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js:
```diff
@@ -0,0 +1,1389 @@
+  MessageLong.setCallback("[EffectRack1_EffectUnit1", "!effectknob3", this.effectKnob);
+
+  MessageLong.addControl("[EffectRack1_EffectUnit2]", "mix", 0x1F, "H");
+  MessageLong.addControl("[EffectRack1_EffectUnit2", "!effectknob1", 0x21, "H");
```
same here, forgot the closing bracket?

Be-ing replied:
That was intentional to make the string manipulation a bit easier. I'll add a comment.

Swiftb0y commented on res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js:
```diff
@@ -0,0 +1,1389 @@
+}
+
+TraktorS2MK2.registerOutputPackets = function() {
+  OutputTop = new HIDPacket("outputTop", 0x80);
+  OutputBottom = new HIDPacket("outputBottom", 0x81);
```
`var`/`const`?

Swiftb0y reviewed (changes requested):
was only able to look at half of the code right now but I've got some first comments.
Also, can you document the packet layout in a language neutral format (best place would be the wiki page I guess) as well so we can make of it use later with the overhauled engine and HID handling?

Be-ing committed:
NI Traktor Kontrol S2 Mk2: toggle quantize with shift + headphones

Be-ing committed:
NI Traktor Kontrol S2 Mk2: add comment explaining lack of "]"

Be-ing committed:
NI Traktor Kontrol S2 Mk2: declare variables with var

Be-ing commented:
>  Also, can you document the packet layout in a language neutral format (best place would be the wiki page I guess) as well so we can make of it use later with the overhauled engine and HID handling?

https://github.com/mixxxdj/mixxx/wiki/Native-Instruments-Traktor-Kontrol-S2-Mk2#hid-protocol

Though I'm not sure what the uses of putting this on the wiki is. I am not going to rewrite the mapping without access to the hardware to test.

Be-ing committed:
common-hid-packet-parser.js: allow setting callback with addControl

This lets scripts register inputs with less boilerplate.

Be-ing committed:
NI Traktor Kontrol S2 Mk2: register input callbacks with addControl

Swiftb0y commented:
> Though I'm not sure what the uses of putting this on the wiki is. I am not going to rewrite the mapping without access to the hardware to test.

Right, that's understandable, but it seems like the controller is pretty popular so there is the chance that someone else will want to reimplement this mapping. This documentation would be a major timesaver for them. Thanks.

ywwg commented:
reviewing now

ywwg commented on res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js:
```diff
@@ -0,0 +1,1336 @@
+  this.last_tick_time = [0.0, 0.0];
+  this.sync_enabled_time = {};
+  
+  this.longPressTimeout = 275; // milliseconds
```
call this this.longPresTimeoutMsec

ywwg commented on res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js:
```diff
@@ -0,0 +1,1336 @@
+  // Values in the short message are all buttons, except the jog wheels.
+  // An exclamation point indicates a specially-handled function.  Everything else is a standard
+  // Mixxx control object name.
+  // "[Channel1]" and "[Channel2]" refer to the left deck or right deck, and may be Channel1 or 3 depending
```
Let's keep this "deck1" and "deck2" (consistent with the common lib) or "[ChannelA]" "[ChannelB]" so it's less confusing.

Be-ing replied:
This is only a 2 deck controller though

ywwg replied:
ok then remove confusing comment

ywwg commented on res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js:
```diff
@@ -0,0 +1,1336 @@
+  MessageShort.addControl("[Channel1]", "!shift", 0x0B, "B", 0x08, false, this.shift);
+  MessageShort.addControl("[Channel1]", "!sync_enabled", 0x0B, "B", 0x04, false, this.syncEnabled);
+  MessageShort.addControl("[Channel1]", "!cue_default", 0x0B, "B", 0x02, false, this.cue);
+  MessageShort.addControl("[Channel1]", "!play", 0x0B, "B", 0x01, false, this.play);
```
Please make the handlers playButton or handlePlay for consistency.  Same with other handlers.

ywwg commented on res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js:
```diff
@@ -0,0 +1,1336 @@
+  MessageShort.addControl("[Channel1]", "!load_track", 0x0C, "B", 0x08, false, this.loadTrack);
+  MessageShort.addControl("[EffectRack1_EffectUnit1]", "!effect_focus_button",
+    0x0E, "B", 0x10, false, this.effectFocusButton);
+  // closing "]" intentionally left off here to make string manipulation a bit easier
```
This is has a slightly  bad Code Smell, where does this make things easier?

ywwg commented on res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js:
```diff
@@ -0,0 +1,1336 @@
+  MessageLong.addControl("[EffectRack1_EffectUnit2", "!effectknob2", 0x23, "H", null, false, this.effectKnob);
+  MessageLong.addControl("[EffectRack1_EffectUnit2", "!effectknob3", 0x25, "H", null, false, this.effectKnob);
+  
+  for (var i = 1; i < 3; i++) {
```
for loops that start at 1, can the end parameter be i <= 2? I find that a little more easy to understand.  (optional)

Be-ing replied:
I find that harder to understand personally.

ywwg commented on res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js:
```diff
@@ -0,0 +1,1336 @@
+
+  // The master gain knob controls the internal sound card volume, so if this was mapped
+  // the gain would be double-applied.
+  //MessageLong.addControl("[Master]", "volume", 0x11, "H");
```
Is this also true for your controller?

Be-ing replied:
yes

ywwg commented on res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js:
```diff
@@ -0,0 +1,1336 @@
+}
+
+TraktorS2MK2.registerOutputPackets = function() {
+  var OutputTop = new HIDPacket("outputTop", 0x80);
```
note that if we merge #2971 you'll have to update the HIDPacket constructors

ywwg commented on res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js:
```diff
@@ -0,0 +1,1336 @@
+
+TraktorS2MK2.linkDeckOutputs = function(key, callback) {
+  // Linking outputs is a little tricky because the library doesn't quite do what I want.  But this
+  // method works.
```
these links look fine, is the comment out of date?

Be-ing replied:
I think you wrote that comment for the S4 Mk2 mapping :P

ywwg replied:
Yeah maybe the code has been updated since then?  :P

ywwg commented on res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js:
```diff
@@ -0,0 +1,1336 @@
+
+TraktorS2MK2.lightDeck = function(group) {
+  // Freeze the lights while we do this update so we don't spam HID.
+  this.freeze_lights = true;
```
batching, etc.

ywwg commented on res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js:
```diff
@@ -0,0 +1,1336 @@
+TraktorS2MK2.init = function(id) {
+  TraktorS2MK2.registerInputPackets();
+  TraktorS2MK2.registerOutputPackets();
+//   var data = [0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x00, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f, 0x7f];
```
remove debug or move it to a debugLights function

Be-ing replied:
I deleted the debugLights function because that didn't work and it was easier to write this than figure out why debugLights didn't work.

ywwg replied:
ok but what I'm saying is, either delete this or create a new debugLights based on this snippet.  I don't want to have commented-out stuff in the committed file

Be-ing replied:
I'd rather have commented out code than a function that confusingly isn't used.

ywwg replied:
how about:

```
DEBUG_LIGHTS = false

debugLights = function() { } 

init = function() {
  ...
  if (DEBUG_LIGHTS) {
    debugLights();
  }
  ...
}
```

ywwg replied:
(obviously pseudojs)

Be-ing replied:
That would be more cumbersome to use. As it is, I only have to uncomment these lines and comment the call to `TraktorS2MK2.registerOutputPackets()` right above it.

ywwg replied:
In my fix, you only have to change false to true to enable the debugging, so the usage is not cumbersome.  We need to treat our javascript more like real code, because it is real code. We've made some progress but we need to make sure that our attitude for controller scripts is more like our attitude for C++. That includes fixing linter errors and not leaving debug code sprinkled around, commented. Please fix this as I described.

Be-ing replied:
https://github.com/mixxxdj/mixxx/pull/2179#issuecomment-505642498

ywwg commented on res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js:
```diff
@@ -0,0 +1,1336 @@
+}
+
+// There are no buttons handled by the long message, so this is a little simpler.  (Even though
+// this is very similar to the other handler, it's easier to keep them separate to know what's
```
which other handler? Be more specific please.

ywwg commented on res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js:
```diff
@@ -0,0 +1,1336 @@
+
+TraktorS2MK2.shift = function(field) {
+  var group = field.id.split(".")[0];
+  TraktorS2MK2.shift_pressed[group] = field.value;
```
the common library has support for shift buttons!

Be-ing replied:
I would rather not rely on the library's mysterious magic.

ywwg replied:
the best way to improve a library is to use it and then update/refactor/improve it as needed.  Regardless, I was wrong, there isn't shift support anyway

ywwg commented on res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js:
```diff
@@ -0,0 +1,1336 @@
+        return;
+      }
+      // was enabled, and button has been let go.  maybe latch it.
+      if (now - TraktorS2MK2.sync_enabled_time[field.group] > 300) {
```
this shouldn't be necessary I think. Just bind value > 0 to sync_enabled and the latching should work without any hacks.

Be-ing replied:
That's not how it works, but I think that's how it should work...

ywwg replied:
as I pointed out on other pull requests, controller scripts shouldn't change mixxx's behavior due to personal preferences of script developers.  We should keep the bindings consistent.  Then, when we change how the sync button works (if we want to remove latching) then all the controller scripts will get the update.  For now, just bind to sync_enabled and don't use your own latching code, or else we'll be stuck with it! (we will have to decide how to bind momentary sync, auto sync, and explicit master for 2.4)

Be-ing replied:
> controller scripts shouldn't change mixxx's behavior due to personal preferences of script developers

I agree, which is why I'm leaving it as is.

>  For now, just bind to sync_enabled and don't use your own latching code, or else we'll be stuck with it! (we will have to decide how to bind momentary sync, auto sync, and explicit master for 2.4)

As I said above, that doesn't work as you think it does and it never has.

ywwg replied:
on the vci-400, the sync buttons are bound like this:

```
<control>
   <group>[Channel1]</group>
   <key>sync_enabled</key>
   <description>MIDI Learned from 22 messages.</description>
   <status>0x92</status>
   <midino>0x01</midino>
   <options>
      <normal/>
   </options>
</control>
```

Be-ing replied:
This is an HID controller.

ywwg replied:
Ah, the problem is that the latching effect was implemented as part of the midi button handler code, not the controlobject code.  So that's why it doesn't work over HID by setting the control object explicitly.

ywwg commented on res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js:
```diff
@@ -0,0 +1,1336 @@
+    } else if (ShiftCueButtonAction == "REVERSEROLL") {
+      engine.setValue(field.group, "reverseroll", field.value);
+    } else {
+      print ("Traktor S2 Mk2 WARNING: Invalid ShiftCueButtonAction picked.  Must be either REWIND " +
```
this should be checked on init and throw an error when the script is loaded.

Be-ing replied:
It would be easy to miss the warning that way. As it is, it is easy to find the warning in the log when the button is pressed.

ywwg replied:
but it would pop up the alert right away -- I don't get how that would be missed.

Be-ing replied:
ah, I forgot we can now throw errors and have them show a popup

Be-ing replied:
done in 3df3b3a

ywwg commented on res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js:
```diff
@@ -0,0 +1,1336 @@
+  }
+}
+
+TraktorS2MK2.jogTouch = function(field) {
```
add a TODO to move this to a common library

Be-ing replied:
I'm not sure I agree with that.

ywwg replied:
this is going to be the third controller script that copies and pastes this jogwheel handling code.  That sort of duplication is unnecessary and brittle

Be-ing replied:
I'm not investing more energy into this legacy API that we will be replacing soon.

Be-ing replied:
https://github.com/mixxxdj/mixxx/wiki/New-controller-system

ywwg commented on res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js:
```diff
@@ -0,0 +1,1336 @@
+  return script.absoluteLin(value, 0, 1, 16, 4080);
+}
+// Tell the HIDController script to use setParameter instead of setValue.
+TraktorS2MK2.scalerParameter.useSetParameter = true;
```
should this be at the top of the file?

Be-ing replied:
No, TraktorS2MK2.scalerParameter is defined right above this.

ywwg replied:
Why is this done outside of the functions, though? it looks like it should happen in init()

Be-ing replied:
67fbe3d27dc97e6fc280c5626e8016a5f67a0e64

ywwg commented on res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js:
```diff
@@ -0,0 +1,1336 @@
+}
+
+TraktorS2MK2.onVuMeterChanged = function(value, group, key) {
+  // This handler is called a lot so it should be as fast as possible.
```
TODO: we should also make this generic, vu meters are super common

Be-ing replied:
How could this be generic? Every controller implements communication with meter LEDs differently.

ywwg replied:
function(volume, num_segments) {
return [an array of values 0-255]
}

ywwg commented on res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js:
```diff
@@ -0,0 +1,1336 @@
+  TraktorS2MK2.current_pad_mode[group] = padMode;
+}
+
+TraktorS2MK2.pad = function(field) {
```
this function is quite long -- can you split it into three, one that handles samplers, one that handles hotcues, and one that decides which to call?

Be-ing replied:
good idea

ywwg commented on res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js:
```diff
@@ -0,0 +1,1336 @@
+  var padMode = TraktorS2MK2.current_pad_mode[field.group];
+  if (padMode !== TraktorS2MK2.pad_modes.intro_outro) {
+    TraktorS2MK2.setPadMode(field.group, TraktorS2MK2.pad_modes.intro_outro);
+    TraktorS2MK2.controller.setOutput(field.group, "!flux_button", 0x7f, false);
```
what is the flux button supposed to do? I have one on my S3 and I don't get it.  Also too, what's a "grid" button?

Be-ing replied:
This is what flux is supposed to do: https://bugs.launchpad.net/mixxx/+bug/1475303

ywwg commented on res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js:
```diff
@@ -0,0 +1,1336 @@
+    };
+  };
+  
+  // FIXME: Why do the LEDs flicker?
```
sometimes the controller wants to do the lights itself -- that was the case on my VCI400

Be-ing replied:
I think there's a subtle issue with the code, but I haven't figured it out.

ywwg commented on res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js:
```diff
@@ -0,0 +1,1336 @@
+    var effectGroup;
+    var key;
+    if (focusedEffect === 0) {
+      effectGroup = effectUnitGroup.slice(0, -1) + "_Effect" + (i+1) + "]";
```
I thnk this is the string manipulation you mentioned before.  Instead, you could just do a simple regex instead (regexes are very fast) https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace and https://opnsrce.github.io/javascript-performance-tip-precompile-your-regular-expressions

ywwg commented on res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js:
```diff
@@ -0,0 +1,1336 @@
+  }
+}
+
+TraktorS2MK2.effectFocusButton = function(field) {
```
can you add some comments for all these functions? I'm not sure what they are doing.

Be-ing replied:
https://github.com/mixxxdj/mixxx/wiki/standard-effects-mapping

ywwg replied:
maybe just link to that page?  
// This deck implements the standard effects mapping as described at [link]

Be-ing replied:
done

ywwg commented on res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js:
```diff
@@ -0,0 +1,1336 @@
+}
+
+TraktorS2MK2.leftEncoder = function(field) {
+  // TODO: common-hid-packet-parser looks like it should do deltas, but I can't get them to work.
```
are they used in any other mappings?

Be-ing replied:
Again, I think you wrote this comment :P

ywwg replied:
Have a look just in case it's fixed itself since then? :)

Be-ing replied:
I couldn't get the library to work either and instead factored this code into a common encoder handling function for the controller in 6914d09

ywwg reviewed (changes requested):
Thanks for working on this! Getting modern controllers mapped is a huge issue for Mixxx

Be-ing commented on res/controllers/Traktor Kontrol S2 Mk2.hid.xml:
```diff
@@ -0,0 +1,19 @@
+        <wiki>https://github.com/mixxxdj/mixxx/wiki/Native-Instruments-Traktor-Kontrol-S2-Mk2</wiki>
+        <devices>
+            <product protocol="hid" vendor_id="0x17cc" product_id="0x1320"
+                     usage_page="0x0" usage="0x0" interface_number="0x3" />
```
I hacked hidcontroller.cpp to print to the log what Mixxx was using and copied those to the XML file.

ywwg replied:
you might need to do that on windows. From what I read on the internet, usage_page / usage are windows-only

Be-ing replied:
This XML file works on Windows.

ywwg replied:
does it work without those parameters?  Might as well remove them if they are noise

Be-ing replied:
I don't think this `product` element is used at all. I think it's just documentation in case we implement HID hotplug in future.

Be-ing committed:
common-hid-packet-parser.js: warn if callback is not a function

Be-ing committed:
NI Traktor Kontrol S2 Mk2: rename freeze_lights variable

Be-ing committed:
NI Traktor Kontrol S2 Mk2: rename longPressTimeout variable

Be-ing committed:
NI Traktor Kontrol S2 Mk2: remove obsolete comment

Be-ing committed:
NI Traktor Kontrol S2 Mk2: make function names more consistent

Holzhaus commented:
There are a bunch of eslint warnings. Are you using the pre-commit hooks?

Be-ing committed:
NI Traktor Kontrol S2 Mk2: split padButton into separate functions

Be-ing committed:
NI Traktor Kontrol S2 Mk2: various fixes for LED output

* Add ButtonBrightnessOff & ButtonBrightnessOn user options.
* Light all buttons to ButtonBrightnessOff state on initialization.
* Fix scaling of color values for hotcues
* Scale pad colors by ButtonBrightnessOn

Be-ing commented:
> There are a bunch of eslint warnings. Are you using the pre-commit hooks?

No, I have to use `git commit --no-verify` with this legacy code or I wouldn't get anything done.

Be-ing committed:
NI Traktor Kontrol S2 Mk2: add comments refering to wiki page

Be-ing committed:
NI Traktor Kontrol S2 Mk2: remove unnecessary detail from comment

Be-ing committed:
NI Traktor Kontrol S2 Mk2: use script.effectUnitRegEx

Be-ing committed:
NI Traktor Kontrol S2 Mk2: add encoder delta calculation function

Holzhaus commented:
> > There are a bunch of eslint warnings. Are you using the pre-commit hooks?
> 
> No, I have to use `git commit --no-verify` with this legacy code or I wouldn't get anything done.

No need to use `--no-verify`. You can skip individual hooks, e. g. `SKIP=eslint git commit`, so that the other hooks can still do their work.

Can you please run the eslint hook on this file at least once? It should fix a lot of the issues automatically. Then you can still ignore the remaining issue if you have to.

Be-ing committed:
NI Traktor Kontrol S2 Mk2: change dim LED brightness to 1

9 is too easy to confuse with full lighting in dark environments
with the power supply plugged in.

Be-ing committed:
NI Traktor Kontrol S2 Mk2: automatic eslint fixes

Be-ing committed:
NI Traktor Kontrol S2 Mk2: remove unused function

Be-ing committed:
NI Traktor Kontrol S2 Mk2: use camelCase for variable names

ywwg commented:
> No need to use `--no-verify`. You can skip individual hooks, e. g. `SKIP=eslint git commit`, so that the other hooks can still do their work.

do we have a wiki page on presubmit hooks?  I find it difficult to find documentation

ywwg commented on res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js:
```diff
@@ -0,0 +1,1378 @@
+        "[Channel2]": []
+    };
+
+    this.last_tick_val = [0, 0];
```
it sounds like we want to use camelCase for everything

ywwg commented on res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js:
```diff
@@ -0,0 +1,1378 @@
+};
+
+TraktorS2MK2.samplerGainKnob = function(field) {
+    for (var i = 1; i <= 8; i++) {
```
please choose either "<" or "<=" for i=1 loops so it's consistent

Be-ing replied:
done 4ab004c

ywwg commented on res/controllers/common-hid-packet-parser.js:
```diff
@@ -415,6 +424,14 @@ HIDPacket.prototype.addControl = function(group, name, offset, pack, bitmask, is
     control_group[field.id] = field
+    
+    if (callback !== undefined) {
+        if (typeof callback === "function") {
```
this is a nit, but putting exception clauses first makes for neater code down the line.  I realize my example may not have followed that rule:
```
if (callback === undefined) {
   return;
}
if (typeof callback !== "function") {
  HIDDebug("ERROR..."
  return;
}
this.setCallback(group, name, callback);
```

ywwg reviewed (changes requested):
this is looking good! 

I agree that we should make new files pass eslint even if you're copying and pasting.

Be-ing committed:
NI Traktor Kontrol S2 Mk2: more eslint fixes

Be-ing committed:
NI Traktor Kontrol S2 Mk2: scroll 5 rows with shift + browse

Holzhaus commented:
> > No need to use `--no-verify`. You can skip individual hooks, e. g. `SKIP=eslint git commit`, so that the other hooks can still do their work.
> 
> do we have a wiki page on presubmit hooks? I find it difficult to find documentation

https://pre-commit.com/#temporarily-disabling-hooks

Be-ing committed:
NI Traktor Kontrol S2 Mk2: add eslint global defines

Finally eslint passes

Be-ing committed:
common-hid-packet-parser.js: refactor error handling

Be-ing committed:
NI Traktor Kontrol S2 Mk2: add error handling for user options

Be-ing committed:
NI Traktor Kontrol S2 Mk2: use <= consistently for loop conditions

Be-ing committed:
NI Traktor Kontrol S2 Mk2: rename encoderDelta to encoderDirection

Be-ing committed:
NI Traktor Kontrol S2 Mk2: reduce string manipulation

Be-ing commented:
There are a bunch of ESLint errors for the common-hid-packet-parser.js library which IMO are beyond the scope of this PR, but the script file for the controller now passes ESLint.

Be-ing committed:
NI Traktor Kontrol S2 Mk2: document known issue of LED flickering

Be-ing commented:
Ready for merge?

ywwg commented:
> https://pre-commit.com/#temporarily-disabling-hooks

yes but I mean can we have a mixxx-specific page that talks more about how we specifically use pre-commit and what our policies are. Similar to how we explain how to use scons (soon to be cmake)

Be-ing committed:
NI Traktor Kontrol S2 Mk2: refactor LED debugging

Be-ing committed:
NI Traktor Kontrol S2 Mk2: move setting of useSetParameter

Holzhaus commented:
> > https://pre-commit.com/#temporarily-disabling-hooks
> 
> yes but I mean can we have a mixxx-specific page that talks more about how we specifically use pre-commit and what our policies are. Similar to how we explain how to use scons (soon to be cmake)

https://github.com/mixxxdj/mixxx/wiki/Using-Git#set-up-automatic-code-checking

ywwg commented:
> https://github.com/mixxxdj/mixxx/wiki/Using-Git#set-up-automatic-code-checking

great, thanks.  I was trying to find it in the "compiling for linux" instructions which are out of date anyway.

ywwg commented on res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js:
```diff
@@ -0,0 +1,1381 @@
+    this.lastTickTime = [0.0, 0.0];
+    this.syncEnabledTime = {};
+
+    this.longPressTimeoutMilliseconds = 275; // milliseconds
```
(comment is now redundant)

ywwg reviewed (approved):
this looks good!  thanks for mapping an important deck :)

Be-ing committed:
NI Kontrol S2 Mk2: remove redundant comment

ywwg commented:
If you're done making updates, feel free to merge

Be-ing commented:
Thanks for the prompt review. I added this to the [wiki](https://github.com/mixxxdj/mixxx/wiki/Hardware%20compatibility#mixxx-certified-mappings).
